### PR TITLE
#159456828 Local login with email/username and password

### PR DESCRIPTION
**Pull request by Veraclins**
#### What does this PR do?
Add ability to login a user locally using email or username and password
#### Description of Task to be completed?
- add support for logging in with email and password
- add support for logging in with username and password
- change the status code and message for the fallback route
#### How should this be manually tested?
- Clone the repo: Run `git clone https://github.com/andela/elven-ah.git` in the folder where you want the repo to be saved.
- Navigate into the newly created folder and install the dependencies: `cd elven-ah && npm install`
- Copy .env.example to .env: `cp .env.example .env`
- Provision a PostgreSQL database and add the credentials to the .env file. 
- Start the server using `npm start`
- Make a POST request to http://localhost:3000/api/auth/signup using postman with your email and verify the email sent to your email.
- Make a POST request to http://localhost:3000/api/auth/login with
    - username or email and
    - password

#### What are the relevant pivotal tracker stories? 
[#159456828](https://www.pivotaltracker.com/story/show/159456828)

### CODE REVIEW

File: server/index.js [hunk 1]
```diff
@@ -33,59 +32,25 @@ app.use(
...
-    message: 'Welcome to Author\'s Haven API',
-  });
-});
-
 // catch un-available routes
 app.all('*', (req, res) => {
   res.status(404).json({
     status: 'error',
-    message: 'Route unavailable on this server',
+    message: 'Oh-oh! Seems like the page you requested does not exist. Please check the url again.',
```
Comment (yomigeek) [inline server/index.js:45]: Good one here, I like how descriptive this message is, thanks for using your discretion in taking care of this part.

File: server/index.js [hunk 2]
```diff
@@ -33,59 +32,25 @@ app.use(
...
-// production error handler
-// no stacktraces leaked to user
+// no stack traces leaked to user in production
 app.use((err, req, res) => {
   res.status(err.status || 500);
   res.json({
     errors: {
       message: err.message,
-      error: {},
+      error: env === 'production' ? {} : err,
```
Comment (yomigeek) [inline server/index.js:57]: Good thought @Veraclins

File: server/tests/users.test.js [hunk 3]
```diff
@@ -35,27 +34,3 @@ describe('Creating and verifying JWT tokens', () => {
...
-    chai.request(server)
-      .get('/api/users')
-      .end((err, res) => {
-        expect(res.status).to.be.equal(401);
-        done();
-      });
-  });
-
-  it('it should return error if invalid token is supplied', (done) => {
-    const invalid = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyIjp7ImlkIjozMCwidXNlcm5hbWUiOiJBbmRlbGEiLCJlbWFpbCI6ImFuZGVsYUB0ZXN0cy5jb20ifSwiaWF0IjoxNTMyODc5MDE1LCJleHAiOjE1MzU0NzEwMTV9.y0w4bC53q0DQZh__6G0LNhyFw_HJaeVp2eN1E8YrJO0';
```
Comment (yomigeek) [inline server/tests/users.test.js:51]: Good job @Veraclins . I feel this could be named invalidToken so it can be more descriptive.

**Review by Oluwafayokemi — COMMENTED** [review]
Hi Clinton! nice job
I have a few concern as to why you altered the mock test. api/auth/mock. I think the reason you have 404 is that you are calling api/auth/mocks instead of api/auth/mock.  Do you mind looking into that?

File: server/controllers/AuthController.js [hunk 4]
```diff
@@ -157,6 +173,44 @@ class AuthController {
...
+        const passwordIsValid = bcrypt.compareSync(password, user.password);
+        if (!passwordIsValid) {
+          const errors = { user: ['Incorrect credentials! Please try again'] };
+          return AuthController.failureResponse({ code: 401, errors }, res);
+        }
+
+        const { id, firstName, lastName } = user;
+
+        const status = { code: 200, message: 'login successful!' };
+        AuthController.successResponse(status, {
```
Comment (chukwuemekachm) [inline server/controllers/AuthController.js:207]: Well done on completing this task @Veraclins. I love the fact that you took initiative and tried to optimize the codebase by creating to methods to handle successful and failed authentication. It shows that you are really attentive to details. However, on this line, you added the user's email and username to the user object you sent to the `successResponse` method. I believe the method signs the token, but then are you certain that the said user did log in with a username in lieu of an email? Please correct me if I'm mistaken.

I love that you tested all the edge cases in this code. But don't you feel it would be much better if you also test the error messages along with the success messages as well?

Well done.

**Review by chukwuemekachm — CHANGES_REQUESTED** [review]

File: server/controllers/AuthController.js (hunk 4 above)
Comment (Veraclins) [reply server/controllers/AuthController.js:207]: Thanks, @chukwuemekachm for your feedback. I don't quite get what you mean by adding the password to the user object sent to the `succesResponse` method. As for whether or not the user signed in with username or email, I don't think that matters as the response on successful login should be the same for all users irrespective of how they login.

Comment (chukwuemekachm) [reply server/controllers/AuthController.js:207]: Sorry I made a mistake, I meant email and username. However, I just corrected that.
What I meant is as follows:
- You destructured `email`, `username` and `password` from the request body. It can only be one of either `username` and `email`. I hoped you would reassign them from the user's object on successful login.
- Now if the user logs in with a `username` the `email` would be null and vice-versa.
- You sent those variables along with the user object. Please resolve it.

Comment (Veraclins) [reply server/controllers/AuthController.js:207]: Thanks for that. I now understand your point and I have made the required changes. Thanks for pointing it out!

**Review by chukwuemekachm — APPROVED** [review]

**Review by madeofhuman — APPROVED** [review]
Nice work attending to the changes requested. You've done a really good job finishing this task, Clinton.

**Review by Oluwafayokemi — APPROVED** [review]

**Review by yomigeek — APPROVED** [review]
Good work @Veraclins
